Implement credsStore in credential config files

matejvasek opened the pull request:
Config file on macOS may contain "credsStore" property indicating what cred store should be used.
This patch checks for its presence and uses it if appropriate.

Signed-off-by: Matej Vasek <mvasek@redhat.com>

mtrmac commented:
(Drive-by comment, by no means a review:)
- This is not macOS specific
- Read support without write support can have surprising behavior
- See also #737 (and #971 ).

matejvasek commented:
@QiWang19 does https://github.com/containers/image/pull/971 fixes this?

matejvasek commented:
@vrothberg do you know whether #971 fixes this?

vrothberg commented:
Would https://github.com/containers/image/pull/1193 help?

lkingland commented:
This seems to be working for me.

Given the following `~/.docker/config.json`
```
{
  "HttpHeaders" : {
    "User-Agent" : "Docker-Client/19.03.12 (darwin)"
  },
  "credsStore" : "desktop",
  "auths" : {
    "https://index.docker.io/v1/" : {

    }
  },
  "experimental" : "enabled",
  "stackOrchestrator" : "swarm"
}
```
The following program:
```
package main

import (
        "fmt"
        "github.com/containers/image/v5/pkg/docker/config"
        "log"
)

func main() {
        dauth, err := config.GetAllCredentials(nil)
        if err != nil {
                log.Fatal(err)
        }
        fmt.Printf("dauths: %+v\n", dauth)
}
```
accesses credentials from the credsStore:
```
> go run main.go
dauths: map[index.docker.io:{Username:[REDACTED] Password:[REDACTED] IdentityToken:}]
```

lkingland commented:
> Would #1193 help?

Unfortunately, when trying with #1193, the Username and Password output are blank

rhuss commented:
I have a similar issue, in that `config.GetAllCredentials(nil)` does not find my real passwords, but returns bogus (possibly stale) passwords. Even when I do a `docker logout` the passwords are still present.

I'm a bit puzzled as I checked all credentials helpers available (docker-credential-desktop,  docker-credential-ecr-login, docker-credential-osxkeychain) and none has the bogus password that `containers/images` find. 

My config.json looks the same:

```json
{
	"auths": {
		"index.docker.io": {},
		"public.ecr.aws": {},
		"quay.io": {}
	},
	"credsStore": "desktop",
	"experimental": "enabled",
	"stackOrchestrator": "swarm"
}%
```

Any help is highly appreciated.

matejvasek commented:
@vrothberg ^^^ any idea what could be issue for Roland?

mtrmac commented:
@rhuss this seems to have nothing to do with this proposed PR, please file a separate issue.

rhuss commented:
> @rhuss this seems to have nothing to do with this proposed PR, please file a separate issue.

I added my comments here, as I think that is the issue for implementing support for credential helpers, right ? --> https://github.com/containers/image/issues/111#issuecomment-861238976

mtrmac commented:
> > Would #1193 help?
> 
> Unfortunately, when trying with #1193, the Username and Password output from the aforementioned program are blank

#1193 requires the configuration to be present in a `registries.conf` option named `credential-helpers`, not in `credsStore`.

rhatdan commented:
@matejvasek Are you still working on this?

matejvasek commented:
@rhatdan for now I made a workaround in our SW. I could get back to this if there is no one else willing to work on this. For now my priority is to fix https://github.com/containers/podman/issues/10776.

matejvasek committed:
Use credstore form docker config

Employ "CredsStore" field of standard docker config.

Signed-off-by: Matej Vasek <mvasek@redhat.com>

matejvasek committed:
Fix test reading non-testing config

The test was reading "~/.docker/config" which can cause test failure.

Signed-off-by: Matej Vasek <mvasek@redhat.com>

matejvasek committed:
Add test for credStore

Signed-off-by: Matej Vasek <mvasek@redhat.com>

matejvasek commented:
@mtrmac @vrothberg PTAL

matejvasek committed:
Fix style

Signed-off-by: Matej Vasek <mvasek@redhat.com>

mtrmac commented on pkg/docker/config/config.go:
```diff
@@ -440,6 +451,14 @@ func RemoveAllAuthentication(sys *types.SystemContext) error {
+				if auths.CredsStore != "" {
+					if creds, err := listAuthsFromCredHelper(auths.CredsStore); err == nil {
+						for registry := range creds {
+							_ = deleteAuthFromCredHelper(auths.CredsStore, registry)
```
Same here.

mtrmac commented on pkg/docker/config/config.go:
```diff
@@ -646,6 +665,12 @@ func findAuthentication(ref reference.Named, registry, path string, legacyFormat
 	}
 
+	if auths.CredsStore != "" {
+		if cred, err := getAuthFromCredHelper(auths.CredsStore, registry); err == nil {
```
Why is it OK to completely ignore errors here?

mtrmac commented on pkg/docker/config/config_test.go:
```diff
@@ -559,6 +559,7 @@ func TestGetAuthFailsOnBadInput(t *testing.T) {
 }
 
 func TestGetAllCredentials(t *testing.T) {
```
Please add tests for the other modified functions as well.

matejvasek replied:
>Please add tests for the other modified functions as well.

Are those functions properly tested right now anyway? I believe they are not tested with respect to helpers right now.

matejvasek replied:
>I believe they are not tested with respect to helpers right now.

For instance the mock helper does noting for the store operation so you can really assert it written anything when calling the `SetCredentials` function.

mtrmac commented on pkg/docker/config/config_test.go:
```diff
@@ -636,6 +637,42 @@ func TestGetAllCredentials(t *testing.T) {
+	require.NoError(t, err)
+	sys = types.SystemContext{
+		AuthFilePath:                authFilePath,
+		SystemRegistriesConfPath:    "",
+		SystemRegistriesConfDirPath: "",
```
This is (sadly?) the same as leaving the values unset, in particular it doesn’t disable any ordinary users’ configuration. So if this data matters, it should point at real files.

matejvasek replied:
@mtrmac Do I need to set it? I think that for this test `AuthFilePath` is enough, isn't it?

mtrmac commented on pkg/docker/config/config_test.go:
```diff
@@ -636,6 +637,42 @@ func TestGetAllCredentials(t *testing.T) {
+	os.Setenv("HOME", dir)
+	return func() {
+		if oldHomeExists {
+			os.Setenv("HOME", oldHome)
```
I don’t immediately see that this is necessary (the existing `GetAllCredentials` test got by without), and anyway, AFAIK it doesn’t work; `containers/storage/pkg/homedir.Get()` caches the value throughout its lifetime, so a single-test override is either completely ineffective, or is effective only if that test happens to run first.

So, to make some of the code testable, we had to introduce `getCredentialsWithHomeDir` and I suspect something similar would be necessary in other cases that tried to override `HOME`.

matejvasek replied:
@mtrmac without this the test is failing on my machine because the  `GetAllCredentials` function also returns credentials from my docker config.

matejvasek replied:
I you don't have `$HOME/.docker/config.json` on your machine it probably works. But for me it's failing.

matejvasek replied:
>caches the value throughout its lifetime, so a single-test override is either completely ineffective, or is effective only if that test happens to run first.

I don't what to do about this.

matejvasek replied:
The test is touch not only testing config files but also actual user config files, that's not good.

mtrmac commented on pkg/docker/config/config.go:
```diff
@@ -440,6 +451,14 @@ func RemoveAllAuthentication(sys *types.SystemContext) error {
 				}
 				auths.CredHelpers = make(map[string]string)
 				auths.AuthConfigs = make(map[string]dockerAuthConfig)
+				if auths.CredsStore != "" {
```
`RemoveAuthentication` also needs adding the support.

mtrmac commented on pkg/docker/config/config.go:
```diff
@@ -85,6 +86,9 @@ func SetCredentials(sys *types.SystemContext, key, username, password string) (s
 				}
+				if auths.CredsStore != "" {
+					return false, setAuthToCredHelper(auths.CredsStore, key, username, password)
+				}
```
Please keep the order of handling `CredHelpers`/`CredsStore`/`AuthConfigs` consistent throughout the package, to make it easier to check that all operations handle the same data. (I guess in the order above, unless there’s a specific reason to do something else.)

mtrmac reviewed (commented):
Thanks.

The implementation looks reasonable reading each operation _in isolation_, but does this actually _work_ end-to-end?

- The user sets `$XDG_RUNTIME_DIR/containers/auth.json` to contain a `credsStore`; that works fine for a while, but is wiped on a reboot. Not good.
- The user sets `~/.docker/config.json` to contain a `credsStore`; that allows reading fine, and survives reboot, but the first modification (`SetCredentials`/`RemoveAuthentication` _only_ reads `auth.json`) _only_ reads `…/auth.json`, doesn’t see any `credsStore` entry in there, and happily writes credentials to `…/auth.json`. That’s not good either.

(Yes, this was already broken with `credHelpers`; in that case it’s more plausible to argue that the read-only path is good enough because the helper might point at a cloud-specific helper and credentials _for that registry_ are never updated via this package. That’s not the case if this is intended to handle all credentials.

I realize that this is not an issue on macOS, where we don’t have `$XDG_RUNTIME_DIR` and the configuration is persistent; having a macOS-only feature is somewhat a stretch, having it available and badly broken on Linux is not really acceptable.

[That’s why we have ended up implementing the `registries.conf` helper configuration.])

@vrothberg PTAL independently, this is tricky and I could well be missing something.

vrothberg commented on pkg/docker/config/config.go:
```diff
@@ -59,7 +60,7 @@ var (
 // NOTE: The return value is only intended to be read by humans; its form is not an API,
 // it may change (or new forms can be added) any time.
-func SetCredentials(sys *types.SystemContext, key, username, password string) (string, error) {
+func SetCredentials(sys *types.SystemContext,  key, username, password string) (string, error) {
```
Nit: not sure why the extra space is needed.

vrothberg reviewed (commented):
Really just found a mini nit in addition to @mtrmac's comments.

matejvasek commented:
>I realize that this is not an issue on macOS, where we don’t have $XDG_RUNTIME_DIR and the configuration is persistent; having a macOS-only feature is somewhat a stretch, having it available and badly broken on Linux is not really acceptable.

@mtrmac actually this is not just macOS specific, you can use credStore also on Linux (and Windows).

matejvasek commented:
>The user sets $XDG_RUNTIME_DIR/containers/auth.json to contain a credsStore; that works fine for a while, but is wiped on a reboot. Not good.

@mtrmac  so user has to login after each reboot with `podman`?

matejvasek commented:
>The user sets $XDG_RUNTIME_DIR/containers/auth.json to contain a credsStore; that works fine for a while, but is wiped on a reboot. Not good.

Doesn't this also apply to already used `credHelpers` field?

matejvasek commented:
Aren't most objections against `credHelpers` also valid to already implemented `credHelpers`?

matejvasek commented:
What about if we implemented only read only part?

matejvasek committed:
revert withTmpHome

The function didn't work as expected, it only worked if the test using
it was run as the first test since HOME envvar is cached somewhere.

Signed-off-by: Matej Vasek <mvasek@redhat.com>

matejvasek committed:
fixup

Signed-off-by: Matej Vasek <mvasek@redhat.com>

matejvasek commented:
@mtrmac What about if we implemented only read only part?

matejvasek commented:
@mtrmac why is `containers/auth.json` not persistent?

matejvasek committed:
add tests

Signed-off-by: Matej Vasek <mvasek@redhat.com>

matejvasek committed:
more error handling

Signed-off-by: Matej Vasek <mvasek@redhat.com>

jankaluza commented:
Hi, and thank you for your contribution!

We’ve recently migrated this repository into a new **monorepo**: [containers/container-libs](https://github.com/containers/container-libs) along with other repositories

As part of this migration, this repository is no **longer accepting new Pull-Requests** and therefore this Pull-Request is being closed.

Thank you very much for your contribution. We would appreciate your continued help in migrating this PR to the new [container-libs repository](https://github.com/containers/container-libs). Please let us know if you are facing any issues.

You can read more about the migration and the reasoning behind it in our blog post: [Upcoming migration of three containers repositories to monorepo](https://blog.podman.io/2025/08/upcoming-migration-of-three-containers-repositories-to-monorepo/).

Thanks again for your work and for supporting the containers ecosystem!
